ethcore/sync/src/chain/supplier.rs

```diff
@@ -150,6 +150,9 @@ impl SyncSupplier {
 
 	/// Respond to GetBlockHeaders request
 	fn return_block_headers(io: &dyn SyncIo, r: &Rlp, peer_id: PeerId) -> RlpResponseResult {
+		// Wait if blocks with forks are being processed
+		// This call will be waiting on client's import mutex, if there are blocks with possible fork in queue
+		io.chain().process_fork();
```

Comment:
Isn't there a way to actually schedule that task after import is completed instead of blocking a thread here? I can imagine a targeted queries can cause the entire node to stall, waiting for block imports, but having no threads to import those blocks?

Reply:
I was thinking about it. There is one problem, that prevents that. We need to send something in response here, if we schedule this request. So the problem is the following: what to send? Empty response is treated by other peers as erroneous. The only sensible idea, that I've got, is to send a special packet "Node is busy (processing fork)", but that would require not obvious protocol changes and won't be supported backwardly.

Reply:
Can't we send nothing and only send the response later? I though the networking is async, so we don't really need to reply synchronously.

Reply:
Yes, I think, we can. I will need to rework https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/chain/supplier.rs#L383 a little bit, in order to implement postponed response.